internal/response-consumer/handler.go

```diff
@@ -214,8 +216,17 @@ func satUpdateRecord(ctx context.Context, tx *gorm.DB, responseFull bool, toUpda
 }
 
 func createRecord(ctx context.Context, tx *gorm.DB, toCreate []db.RunHost) error {
+
+	successOrFailure := clause.OrConditions{Exprs: []clause.Expression{
+		clause.Eq{Column: "run_hosts.status", Value: db.RunStatusSuccess},
+		clause.Eq{Column: "run_hosts.status", Value: db.RunStatusFailure},
+	}}
+
+	notMarkedAsComplete := clause.Where{Exprs: []clause.Expression{clause.Not(successOrFailure)}}
+
 	createResult := tx.Model(db.RunHost{}).
 		Clauses(clause.OnConflict{
+			Where:     notMarkedAsComplete,
 			Columns:   []clause.Column{{Name: "run_id"}, {Name: "host"}},
 			DoUpdates: clause.AssignmentColumns([]string{"status", "log"}),
```

Comment:
@dehort Looking good! Any idea if this `notMarkedAsComplete` check might be redundant, since we are checking for it through the where clause in the main function?

Also, any idea if we need to do the same for satellite requests? I think satellite payloads have a `sat_sequence` number, which kind of protects us against out of order messages, but might be good to double check. :thinking:

Reply:
I believe the `notMarkedAsComplete`  where clause is needed here because we are only passing the transaction object to the `createRecord()` method.  We don't actually pass the query object that we built in the calling function.

Yeah, the `sat_sequence` field protect us from out of order messages at the host level.  But... I think the change in this PR protects us from marking a top level run as `complete` or `failed` and then marking it as `running` if we received a satellite message out of order.  I added another test for this.

Thank you, @tahmidefaz !

Reply:
Sounds good, thank you!